moveit_planners/ompl/ompl_interface/src/detail/state_validity_checker.cpp

```diff
@@ -194,4 +196,112 @@ double ompl_interface::StateValidityChecker::clearance(const ompl::base::State*
   collision_detection::CollisionResult res;
   planning_context_->getPlanningScene()->checkCollision(collision_request_with_distance_, res, *robot_state);
   return res.collision ? 0.0 : (res.distance < 0.0 ? std::numeric_limits<double>::infinity() : res.distance);
+}
+
+/*******************************************
+ * Constrained Planning StateValidityChecker
+ * *****************************************/
+
+bool ompl_interface::ConstrainedPlanningStateValidityChecker::isValid(const ompl::base::State* wrapped_state,
+                                                                      bool verbose) const
+{
+  // Unwrap the state from a ConstrainedStateSpace::StateType
+  auto state = wrapped_state->as<ompl::base::ConstrainedStateSpace::StateType>()->getState();
+
+  // Use cached validity if it is available
+  if (state->as<ModelBasedStateSpace::StateType>()->isValidityKnown())
+    return state->as<ModelBasedStateSpace::StateType>()->isMarkedValid();
+
+  // do not use the unwrapped state here, as satisfiesBounds expects a state of type ConstrainedStateSpace::StateType
+  if (!si_->satisfiesBounds(wrapped_state))
+  {
+    if (verbose)
+      ROS_INFO_NAMED(LOGNAME, "State outside bounds");
+    const_cast<ob::State*>(state)->as<ModelBasedStateSpace::StateType>()->markInvalid();
+    return false;
+  }
+
+  moveit::core::RobotState* robot_state = tss_.getStateStorage();
+  // do not use the unwrapped state here, as copyToRobotState expects a state of type ConstrainedStateSpace::StateType
+  planning_context_->getOMPLStateSpace()->copyToRobotState(*robot_state, wrapped_state);
+
+  // check path constraints
+  const kinematic_constraints::KinematicConstraintSetPtr& kset = planning_context_->getPathConstraints();
+  if (kset && !kset->decide(*robot_state, verbose).satisfied)
+  {
+    const_cast<ob::State*>(state)->as<ModelBasedStateSpace::StateType>()->markInvalid();
```

Comment:
I understand that this an (arguably less ideal) way to implement caching of the validity. I added a single comment at the top of the file to avoid copying the same comment many times:
```C++
/** A state validity checker checks:
 *
 * - Bounds (joint limits).
 * - Collision.
 * - Kinematic path constraints.
 * - Generic user-specified feasibility using the `isStateFeasible` of the planning scene.
 *
 * IMPORTANT: Although the isValid method takes the state as `const ompl::base::State* state`,
 * it uses const_cast to modify the validity of the state with `markInvalid` and `markValid` for caching.
 * **/
```

Reply:
We could also declare the `flags` member of `ModelBasedStateSpace::StateType` as `mutable`, so that it can be changed from a `const` qualified member function. I'm happy to provide a PR.
But one question (as I don't know the details): is the state a shared object between threads? If so, the `flags` should probably be an `atomic` to avoid data races.